### DISCUSSION TITLE
Tasks view: change state a un estados "cerrado"

### DISCUSSION THREAD

**Discussion by ValeMP** [Cotalker/documentation#43] [2. Questions and Answers] [answered]
Hola!

He estado trabajanado en una solcuión que dentro de la maquina de estado (SM padre), la compañia (activo) puede cambiar de abierto a cerrado. Pero hoy cuando estaba probando el cambio, no me aparece la opción! Me podrían explicar un poquito por qué ya no funciona ): para ver que podría hacer. 

Subo fotos de lo que me refiero:
(1) SM 
![Screen Shot 2021-04-06 at 18 36 44](https://user-images.githubusercontent.com/42095617/113786402-0dfa3a00-9707-11eb-93ee-614d2f481d06.png)

(2) Task view
![Screen Shot 2021-04-06 at 18 31 35](https://user-images.githubusercontent.com/42095617/113786427-1a7e9280-9707-11eb-957a-27c2710468ef.png)

Gracias!!

**Comment by rafaelwalter** [comment, marked as answer]
@ValeMP Para cambiar una task de estado abierto a cerrado (de una misma SM) se deben cumplir 2 cosas.

- Pero antes de que veamos estas 2 cosas debemos entender que tasks tienen una estructura de dados de arból. Por qué esto es importante?

![Screen Shot 2021-04-07 at 9 40 53 AM](https://user-images.githubusercontent.com/26169103/113876036-5b1cf100-9785-11eb-80ec-bce47b6435c8.png)

Por que todos los cambios que se hacen en una task tienen efectos secundarios en sus subtasks. En la imagen que tenemos arriba podemos ver que si la task 3 tiene algun cambio, todas sus subtasks (5, 6, 7) podran ser afectadas.

- Cuando queremos cambiar la task 3 de estado tambien aplicamos cambios de estado a las subtasks. Por qué?

Cada estado define su submaquina de estado. Puede que tengan exactamente la misma o diferentes submaquinas de estado. Si los estados de origen y destino tiene la misma submaquina de estado no tendrian ningun efecto. Pero si son submaquinas de estado distintas entonces todas las subtasks pasaran a tener un nuevo estado. 

- Se puede cambiar entre cualquier maquina de estado?

No. Maquinas de estado deben cumplir algunas reglas para tener intercambio de tasks.  (Este es un tema largo y deberia estar en otro topico).  

Ahora que ya entendemos algunos conceptos básicos de tasks podemos ver qué esta afectando el cambio que reportaste:

1. El estado siguiente debe permitir el cambio de submaquinas de estado. Si tienen la misma maquina de estado no deberia ser un problema.

El segundo punto es el que creo que esta afectado tu implementacion:

2. Si una task está cambiando a un estado de tipo `cerrado`, todas sus subtasks tambien deben estar con estados de tipo `cerrados`. Esto no se hace automaticamente ya que es un cambio irreversible. Entonces antes de cambiar una task se debe manualmente cerrar todas sus subtasks.

**Reply by ValeMP** [reply]
okeeei, gracias!
